mlir/lib/Dialect/Rock/Transforms/GemmToGridwise.cpp

```diff
@@ -93,6 +106,132 @@ struct AttentionRewritePattern : public OpConversionPattern<AttentionOp> {
+  Location loc = op->getLoc();
+
+  if (!isa<MemRefType>(a.getType()))
+    return op.emitOpError("Cannot lower unbufferized gemm to gridwise");
+
+  bool isAccel = rock::isAccel(op.getFeatures());
+  if (!isAccel) {
+    return op.emitError("Currently, op is only supported on GPUs "
+                        "with matrix accelerator extentions");
+  }
+  if (!op.getParams0().has_value()) {
+    return op.emitError("gemm0 params is missing and it should've been "
+                        "assigned by affix-tuing-params");
+  }
+  RockAccelTuningParamAttrInterface params0 =
+      cast<RockAccelTuningParamAttrInterface>(op.getParams0Attr());
+  if (!op.getParams1().has_value()) {
+    return op.emitError("gemm1 params is missing and it should've been "
+                        "assigned by affix-tuing-params");
+  }
+  RockAccelTuningParamAttrInterface params1 =
+      cast<RockAccelTuningParamAttrInterface>(op.getParams1Attr());
+
+  // Note: the gridwise ops take K x M and K x N, so A must be transposed if
+  // it's in the natural M x K form
+  a = normalizeMatrix(a, rw, loc, !aTransposed, "gemm0K", "gemm0M");
+  b = normalizeMatrix(b, rw, loc, bTransposed, "gemm0K", "gemm0N");
+  c = normalizeMatrix(c, rw, loc, cTransposed, "gemm1K", "gemm1N");
+  out = normalizeMatrix(out, rw, loc, op.getOTransposed(), "gemm1M", "gemm1N");
+
+  // Note, matrix dimension correctness is handled in the verifier
+  ArrayRef<int64_t> aShape = cast<MemRefType>(a.getType()).getShape();
+  ArrayRef<int64_t> bShape = cast<MemRefType>(b.getType()).getShape();
+  ArrayRef<int64_t> cShape = cast<MemRefType>(c.getType()).getShape();
+  GemmSize gemm0Size(/*g=*/aShape[0], /*m=*/bShape[2],
+                     /*k=*/aShape[1],
+                     /*n=*/aShape[2]);
+  GemmSize gemm0ExtraPad =
+      requiredPadding(params0, gemm0Size).value_or(GemmSize{0, 0, 0, 0});
+  GemmSize gemm1Size(/*g=*/aShape[0], /*m=*/cShape[2],
+                     /*k=*/cShape[1],
+                     /*n=*/aShape[2]);
+  GemmSize gemm1ExtraPad =
+      requiredPadding(params1, gemm1Size).value_or(GemmSize{0, 0, 0, 0});
+
+  a = padMatrix(a, rw, loc, "gemm0K", gemm0ExtraPad.k, "gemm0N",
+                gemm0ExtraPad.n);
+  b = padMatrix(b, rw, loc, "gemm0K", gemm0ExtraPad.k, "gemm0M",
+                gemm0ExtraPad.m);
+  c = padMatrix(c, rw, loc, "gemm1K", gemm1ExtraPad.k, "gemm1M",
+                gemm1ExtraPad.m);
+  // In the transposed layout, from a tuning params point of view
+  // the output dimensions are swapped. Though we will only be
+  // swapping them inside gridwise lowering to keep the surrounding
+  // fusions legit. So the extra pad needs to be swapped and applied.
+  out = padMatrix(out, rw, loc, "gemm1N", gemm1ExtraPad.n, "gemm1M",
+                  gemm1ExtraPad.m);
+
+  if (failed(computeGridSizeAttentionGemmElmtGemm(rw, op, a, b, c))) {
+    return op.emitError(
+        "failed to compute the grid size of `GemmElementwiseGemmOp`");
```

Comment:
```suggestion
        "failed to compute the grid size of `GemmElementwiseGemmOp`");
```

it could also be attention op